File: architecture-skills/object-oriented-programming-paradigm/examples/polymorphism/polymorphism.py
"""Polymorphism in Python.

One loop calls area()/describe() on a mixed list and each object
answers with its own implementation (subtype polymorphism via an ABC).
Python's everyday form is duck typing: Square never inherits Shape,
yet it works at the same call site because it has the right methods.
"""

from __future__ import annotations

import math
from abc import ABC, abstractmethod


class Shape(ABC):
    """The abstraction: the only type the client mentions."""

    @abstractmethod
    def area(self) -> float: ...

    @abstractmethod
    def describe(self) -> str: ...


class Circle(Shape):
    def __init__(self, radius: float) -> None:
        self.radius = radius

    def area(self) -> float:
        return math.pi * self.radius**2

    def describe(self) -> str:
        return f"circle r={self.radius}"


class Rectangle(Shape):
    def __init__(self, width: float, height: float) -> None:
        self.width = width
        self.height = height

    def area(self) -> float:
        return self.width * self.height

    def describe(self) -> str:
        return f"rectangle {self.width}x{self.height}"


class Square:
    """No Shape base class: duck typing (ad-hoc polymorphism).

    Any object with area() and describe() fits the same loop.
    """

    def __init__(self, side: float) -> None:
        self.side = side

    def area(self) -> float:
        return self.side**2

    def describe(self) -> str:
        return f"square {self.side}x{self.side}"


def total_area(shapes) -> float:
    """Parametric in spirit: works for any iterable of shape-like objects."""
    return sum(s.area() for s in shapes)


if __name__ == "__main__":
    shapes = [Circle(1.0), Rectangle(3.0, 4.0), Square(2.0)]

    # One call site, three implementations: the runtime type decides.
    for shape in shapes:
        print(f"{shape.describe():>18}: area {shape.area():.2f}")

    print(f"{'total':>18}: area {total_area(shapes):.2f}")

File: architecture-skills/object-oriented-programming-paradigm/examples/polymorphism/polymorphism.swift
// Polymorphism in Swift.
//
// One loop calls area()/describe() on a mixed list and each object
// answers with its own implementation. Protocols + structs are the
// idiomatic form; Square conforms to Shape explicitly.

import Foundation

protocol Shape {
    func area() -> Double
    func describe() -> String
}

struct Circle: Shape {
    let radius: Double

    func area() -> Double { Double.pi * radius * radius }
    func describe() -> String { "circle r=\(radius)" }
}

struct Rectangle: Shape {
    let width: Double
    let height: Double

    func area() -> Double { width * height }
    func describe() -> String { "rectangle \(width)x\(height)" }
}

struct Square: Shape {
    let side: Double

    func area() -> Double { side * side }
    func describe() -> String { "square \(side)x\(side)" }
}

func totalArea(_ shapes: [Shape]) -> Double {
    shapes.reduce(0) { $0 + $1.area() }
}

let shapes: [Shape] = [Circle(radius: 1.0), Rectangle(width: 3, height: 4), Square(side: 2.0)]

for shape in shapes {
    print(String(format: "%18@: area %.2f", shape.describe(), shape.area()))
}
print(String(format: "%18@: area %.2f", "total", totalArea(shapes)))
